### PR TITLE
Update three marine converters for ioda-converters code sprint

### LINKED ISSUE #894: Upgrade converter for hgodas_insitu2ioda.py, ndbc_hfradar2ioda.py, and  glider2ioda.py to follow the JEDI Data Conventions
Update the converter identified in the title (hereafter referred to as the **target converter**) to follow the JEDI Data Conventions.

- Update variable names
- Ensure variable units match the units in the conventions table
- Update the global attributes dictionary as needed
- Run the converter test to make sure it runs (but will fail because of unupdated test reference)
- Run ```ioda-validate.x``` on your output file and verify that there are no errors or warnings
- Update test reference file
- Run unit tests making use of the target converter and verify that they pass
- Make sure you verify the results by examining the file using ```ncdump -t```, ```h5dump```, ```HDFview```, or a similar application.

## ZenHub organization

Please make sure that you have properly set:
- Labels: OBS3
- Estimate: An estimate of how long this issue will take to resolve. 1 = half a day. 2 = full day. 3 = two days, 5 = three days.
- Epics: [Sprint: Implement the JEDI Data Conventions in the Converters](https://github.com/JCSDA-internal/ioda-converters/issues/813)
- Assignees: Anyone **working** on this issue. Anyone who is not working directly on the issue, but
  who should receive notifications, should be listed below.
- Reviewers

## Notifications

Fill in a list of GitHub handles here. People here are not *assigned*, but instead receive notifications
about this issue's progress.

Notifications should always go to: @rhoneyager.

Other people:

### PR DESCRIPTION
**Pull request by YoulongXia-NOAA**
## Description

The three ioda converters below are upgraded via ioda-converters code sprint:
 
hgodas_insitu2ioda.py 
ndbc_hfradar2ioda.py
glider2ioda.py 

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #894 

## Acceptance Criteria (Definition of Done)

ctest passed and two converters are merged

## Dependencies

No

## Impact

Please list the other repositories (if any) that this PR will require changes in (example below)
No

### CODE REVIEW

**Comment by YoulongXia-NOAA** [comment]
@gthompsnJCSDA, I changed " sea_water_meridional_current" and "sea_water_zonal_current" into "waterMeridionalCurrentSpeed" and "waterZonalCurrentSpeed" right now. What variable names do you suggest? @YanHao-NOAA, do you have any suggestions for them? I googled and did not find similar name although I found currentSpeed. Not sure for u and v in atmosphere for wind speed, how to defined their names? uWind or vWind?

**Comment by YoulongXia-NOAA** [comment]
> @YoulongXia-NOAA I think you need to merge in the staging branch and resolve conflicts. For example, I see this PR tries to remove the IMS snow converter that has already been removed...

Will check and do such a merging soon.

**Comment by YoulongXia-NOAA** [comment]
@YanHao-NOAA, you are the expert for marine. Original name is  " sea_water_meridional_current" and "sea_water_zonal_current". It cannot definitely use "_" for variable names. It is old conventions. We need to use new convention from Greg's table. But neither of your suggested names are there.   Even I search for word "current", nothing is there.  @gthompsnJCSDA,  how about are waterCurrentsEastward and waterCurrentsNorthward? If you look at Yan's suggestions and original names, can you select thyese two variables names?

**Review by travissluka — CHANGES_REQUESTED** [review]
salinity should be unitless, not  `g kg-1`
see comments here https://github.com/JCSDA-internal/ioda-converters/pull/895#pullrequestreview-928943514

**Comment by YoulongXia-NOAA** [comment]
@travissluka, we use unitless or without unit at all. Which one should I use?

**Comment by YoulongXia-NOAA** [comment]
Oh, @travissluka, I saw that it should be unitless. Thank you

**Comment by YoulongXia-NOAA** [comment]
> salinity should be unitless, not `g kg-1` see comments here [#895 (review)](https://github.com/JCSDA-internal/ioda-converters/pull/895#pullrequestreview-928943514)

Changed and re-tested it. Committed now. @travissluka, if you are a marine expert, could you please suggest current speed names I asked? I am a land/atmosphere scientist but very new for marine.

File: src/marine/glider2ioda.py [hunk 1]
```diff
@@ -132,20 +132,21 @@ def main():
...
     VarAttrs = DefaultOrderedDict(lambda: DefaultOrderedDict(dict))
-    VarAttrs[('sea_water_temperature', 'ObsValue')]['units'] = 'celsius'
-    VarAttrs[('sea_water_salinity', 'ObsValue')]['units'] = 'psu'
-    VarAttrs[('sea_water_temperature', 'ObsValue')]['_FillValue'] = -32767
-    VarAttrs[('sea_water_salinity', 'ObsValue')]['_FillValue'] = -32767
-    VarAttrs[('sea_water_temperature', 'PreQC')]['units'] = ''
-    VarAttrs[('sea_water_salinity', 'ObsError')]['units'] = ''
+    VarAttrs[('waterTemperature', 'ObsValue')]['units'] = 'K'
+    VarAttrs[('waterTemperature', 'ObsError')]['units'] = 'K'
+    VarAttrs[('salinity', 'ObsValue')]['units'] = 'unitless'
```
Comment (gthompsnJCSDA) [inline src/marine/glider2ioda.py:144]: This line and one below should be a value of one (1), not unitless.  There is no unit for unitless so any variable without units, this attribute should be skipped entirely.

File: src/marine/hgodas_insitu2ioda.py [hunk 2]
```diff
@@ -25,15 +25,15 @@
 
 vName = {
     #          var name,            units
-    2210: ["sea_water_temperature", "C"],
-    2220: ["sea_water_salinity", "PSU"]
+    2210: ["waterTemperature", "K"],
+    2220: ["salinity", "unitless"]
```
Comment (gthompsnJCSDA) [inline src/marine/hgodas_insitu2ioda.py:29]: Same as earlier comment about units.

File: src/marine/ndbc_hfradar2ioda.py [hunk 3]
```diff
@@ -22,15 +22,14 @@
 from collections import defaultdict, OrderedDict
 from orddicts import DefaultOrderedDict
 
-vName = ["sea_water_meridional_current",
-         "sea_water_zonal_current",
-         "DOPX",
-         "DOPY"]
+vName = ["waterMeridionalCurrentSpeed",
```
Comment (gthompsnJCSDA) [inline src/marine/ndbc_hfradar2ioda.py:25]: I will check the WMO element table, but perhaps we can use ```waterMeridionalVelocity``` and ```waterZonalVelocity```

File: src/marine/ndbc_hfradar2ioda.py [hunk 4]
```diff
@@ -22,15 +22,14 @@
...
+         "waterZonalCurrentSpeed",
+         ]
 
 locKeyList = [
     ("latitude", "float"),
     ("longitude", "float"),
-    ("datetime", "string")
+    ("dateTime", "string")
```
Comment (gthompsnJCSDA) [inline src/marine/ndbc_hfradar2ioda.py:32]: the ```dateTime``` variable should be type ```long```

File: src/marine/ndbc_hfradar2ioda.py [hunk 5]
```diff
@@ -120,18 +118,18 @@ def main():
...
+    DimDict = {'Location': nlocs}
     writer = iconv.IodaWriter(args.output, locKeyList, DimDict)
     VarAttrs = DefaultOrderedDict(lambda: DefaultOrderedDict(dict))
-    VarAttrs[('sea_water_meridional_current', 'ObsValue')]['units'] = 'm/s'
-    VarAttrs[('sea_water_zonal_current', 'ObsValue')]['units'] = 'm/s'
-    VarAttrs[('sea_water_meridional_current', 'ObsValue')]['_FillValue'] = -32767
-    VarAttrs[('sea_water_zonal_current', 'ObsValue')]['_FillValue'] = -32767
-    VarAttrs[('longitude', 'MetaData')]['units'] = 'degree'
-    VarAttrs[('latitude', 'MetaData')]['units'] = 'degree'
+    VarAttrs[('waterMeridionalCurrentSpeed', 'ObsValue')]['units'] = 'm s-1'
```
Comment (gthompsnJCSDA) [inline src/marine/ndbc_hfradar2ioda.py:126]: We should be able to re-use the variable names in list higher up:  ```vname[0]``` and ```vname[1]``` so there is only one spot in the code to change.

**Review by gthompsnJCSDA — CHANGES_REQUESTED** [review]
Some changes are needed.  I can submit the new variable names into the same PR as my other for marine converters (godae_track2ioda.py for example).  That way you just have the ioda PR 673 as a dependency.

**Comment by YoulongXia-NOAA** [comment]
I found speed of current in both WMO table and NCEP GRIB table as speedOfCurrent. I am not sure where eastward and northward should be put? speedOfEastwardCurrent or eastwardSpeedOfCurrent or speedOfCurrentEastward? Any suggestions for this name?

**Comment by YoulongXia-NOAA** [comment]
In NCEP grib table, they are defined uComponentOfCurrent and vComponentOfCurrent. Can we use this definition (https://www.nco.ncep.noaa.gov/pmb/docs/on388/table2.html)? @YanHao-NOAA @gthompsnJCSDA

File: src/marine/ndbc_hfradar2ioda.py (hunk 3 above)
Comment (travissluka) [reply src/marine/ndbc_hfradar2ioda.py:25]: @gthompsnJCSDA those sound fine. I would prefer velocity over speed. Should we have a `...SurfaceVelocity` as well for the same reasons we have the other surface variables? what does the atmosphere do in this case? The case of HF radar here is similar to SST/SSS, whereby the "surface" measurement is some ill defined almost-at-the-surface depth.

File: src/marine/glider2ioda.py (hunk 1 above)
Comment (YoulongXia-NOAA) [reply src/marine/glider2ioda.py:144]: Changed as "1".

File: src/marine/hgodas_insitu2ioda.py (hunk 2 above)
Comment (YoulongXia-NOAA) [reply src/marine/hgodas_insitu2ioda.py:29]: Changed as "1".

File: src/marine/ndbc_hfradar2ioda.py (hunk 3 above)
Comment (YoulongXia-NOAA) [reply src/marine/ndbc_hfradar2ioda.py:25]: Now I used vComponentOfCurrent and uComponentOfCurrent right now. Please check WMO table and let me know your suggestions.

Comment (YoulongXia-NOAA) [reply src/marine/ndbc_hfradar2ioda.py]: @gthompsnJCSDA, @travissluka, in WMO, it has only speed of current and direction of current. But I also found:
SEA/WATER CURRENT SPEED AT A LEVEL
SEA/WATER CURRENT DIRECTION AT A LEVEL
SEA SURFACE CURRENT SPEED
There is no other definitions. Only reference is from NCEP GRIB table definition for x-direction and y direction current definitions.

File: src/marine/hgodas_insitu2ioda.py [hunk 6]
```diff
@@ -65,9 +70,7 @@ def _read(self):
         qcs = ncd.variables['qc'][:]
         ncd.close()
 
-        base_date = datetime(1970, 1, 1) + timedelta(seconds=int(time[0]))
-
-        self.VarAttrs['depth', 'MetaData']['units'] = 'm'
+        dt = qcs.astype('int64')
```
Comment (gthompsnJCSDA) [inline src/marine/hgodas_insitu2ioda.py:73]: I thought I commented on this line before, or perhaps it was in another of the 3 files.  I believe this line is a possible mistake.  The variable ```qcs``` should clearly be filled with QC values and nothing to do with time and I thought the variable called ```dt``` was something related to delta-time.  Please have a closer look at this.

File: src/marine/ndbc_hfradar2ioda.py [hunk 7]
```diff
@@ -117,21 +115,23 @@ def main():
     args = parser.parse_args()
     fdate = datetime.strptime(args.date, '%Y%m%d%H')
     VarDims = {
-        'sea_water_meridional_current': ['nlocs'],
-        'sea_water_zonal_current': ['nlocs']}
+        'vComponentOfCurrent': ['Location'],
```
Comment (gthompsnJCSDA) [inline src/marine/ndbc_hfradar2ioda.py:118]: I am not very fond of these two variable names.  I had already renamed these as:
```
  waterZonalVelocity
  waterMeridionalVelocity
```
as part of my ioda PR 673.  Would you mind changing these to match please and then make your PR depend on 673 in ioda and we could always change our minds later on these variables if someone in ocean group has a problem?

**Review by gthompsnJCSDA — CHANGES_REQUESTED** [review]
Changes are requested.  Getting close to done.

File: src/marine/hgodas_insitu2ioda.py (hunk 6 above)
Comment (YoulongXia-NOAA) [reply src/marine/hgodas_insitu2ioda.py:73]: The purpose for this sentence is to define an array dt, and it is used in below:

**dt[i]** = time[0] + hrs[i]*3600
locKey = lats[i], lons[i], dpth[i], **dt[i]**

It is a correct sentence.

File: src/marine/ndbc_hfradar2ioda.py (hunk 7 above)
Comment (YoulongXia-NOAA) [reply src/marine/ndbc_hfradar2ioda.py]: Changed as suggested. Re-tested and committed.

File: src/marine/hgodas_insitu2ioda.py (hunk 6 above)
Comment (gthompsnJCSDA) [reply src/marine/hgodas_insitu2ioda.py:73]: Ah, I see now that dt is simply being created as a copy of another variable to gain its length and then reassigned later.  Kinda kludgy and I would personally prefer to create an empty and zero filled array using something like this:
```dt = np.full(len(qcs), 0, dtype=np.int64)```

however I will allow the existing solution.

**Comment by srherbener** [comment]
> salinity should be unitless, not `g kg-1` see comments here [#895 (review)](https://github.com/JCSDA-internal/ioda-converters/pull/895#pullrequestreview-928943514)

This has been addressed so I will try the reset on @travissluka's review.

File: src/marine/hgodas_insitu2ioda.py (hunk 6 above)
Comment (srherbener) [reply src/marine/hgodas_insitu2ioda.py:73]: I second @gthompsnJCSDA's suggestion. The code is more clear about what is the intention (create a new variable with a matching length to qcs).

**Review by srherbener — COMMENTED** [review]
Looks good. I just have one comment. Thanks for adding these in!

**Comment by YoulongXia-NOAA** [comment]
> 

Greg uses 1 after travis agreed. I also use 1 @gthompsnJCSDA

**Comment by YoulongXia-NOAA** [comment]
https://github.com/JCSDA-internal/ioda-converters/pull/895#issuecomment-1087941182

**Comment by YoulongXia-NOAA** [comment]
> > salinity should be unitless, not `g kg-1` see comments here [#895 (review)](https://github.com/JCSDA-internal/ioda-converters/pull/895#pullrequestreview-928943514)
> 
> This has been addressed so I will try the reset on @travissluka's review.

Use 1 as Greg used after Travis agreed.

**Comment by YoulongXia-NOAA** [comment]
changed as Greg suggested. Thank you.
================================================================
Youlong Xia, Ph.D
I. M. Systems Group at EMC/NCEP
5830 University Research Court
College Park, MD 20740
Phone: 301-683-3696
=================================================================

File: src/marine/hgodas_insitu2ioda.py (hunk 6 above)
Comment (YoulongXia-NOAA) [reply src/marine/hgodas_insitu2ioda.py]: Now I used waterZonalVelocity and  waterMeridionalVelocity as Greg suggested.

**Comment by gthompsnJCSDA** [comment]
@YoulongXia-NOAA @srherbener  I am trying to run the python scripts to ensure all is working properly, but when I attempt the glider, I get this message...
```
    import yaml
ModuleNotFoundError: No module named 'yaml'
```

So I am no sure why python is not loading the yaml module.  Can anyone help?

**Comment by CoryMartin-NOAA** [comment]
@gthompsnJCSDA the `python/3.9.2` module does not have the yaml library for some reason...

**Comment by gthompsnJCSDA** [comment]
> @gthompsnJCSDA the `python/3.9.2` module does not have the yaml library for some reason...

Thanks Cory.  I will wait to see if ci-testing logs indicate success.  If they all pass, then I will hit the approve button for review.

**Comment by gthompsnJCSDA** [comment]
There was only one tiny item that a blank line in hgodas_insitu contained some space characters.  So I fixed the source code and pushed it to the repo and now I approve the PR because all tests should pass.

**Review by gthompsnJCSDA — APPROVED** [review]
Thank you Youlong for keeping going on this set of 3 converters and making all the suggested changes.  It looks good to me now!

**Comment by YoulongXia-NOAA** [comment]
> Thank you Youlong for keeping going on this set of 3 converters and making all the suggested changes. It looks good to me now!

Thank you @gthompsnJCSDA. I learnt a lot from you and other colleagues. Good experience.

**Comment by YoulongXia-NOAA** [comment]
@srherbener, I think that I have addressed your two comments (Note: salinity unit is 1 rather than unitless after the discussion of Greg and Travis). If your time is available and no further comments, please approve my PR so that I can be merged. Thank you.

**Review by srherbener — APPROVED** [review]
Looks great! Thanks for addressing my comments

**Comment by YoulongXia-NOAA** [comment]
@srherbener, thank you for join in to review my PR. Have a good day.

**Pull request merged**